### ISSUE TITLE
colour balance differences​ between thumbnail and main preview

### ISSUE THREAD

**Issue by agriggio** [Beep6581/RawTherapee#3804]
the code that computes the colour balance (white and black levels, awb,...) for thumbnails is slightly different from that used for the editor window. in some cases, this causes very visible output differences. example and a couple of proposed patches (that probably still need some work) at https://discuss.pixls.us/t/thumbnails-turn-pink/

**Comment by heckflosse** [comment]
@agriggio Alberto, thanks for opening an Issue about that :+1: 
I think, before we open a file in editor we can't get correct Auto WB information without increasing initial processing time of thumb generation. When the thumbs are generated the first time, approximative Auto WB values are written to the cache file.
What we could do, is to update the cache file with correct values when the image is opened first time in editor. What do you think?

**Comment by iliasg** [comment]
Ingo, I think we first have to establish correct use of black level in thumbnails code .. the wrong black level (which also results in wrong white level) is a serious reason for innaccurate WB calculations

**Comment by heckflosse** [comment]
@iliasg I agree. But the wrong black level in thumbnail code deserves an own Issue as well.
btw: I think this is also a good image to test auto WB : http://www.rawtherapee.com/shared/test_images/nikon_d80_wb.nef

**Comment by agriggio** [comment]
@heckflosse @iliasg it's still unclear to me why rtthumbnail uses its own way of determining the colour levels instead of using the code in `rawimagesource`. is there some good reason or is it just a matter of refactoring the code?

**Comment by heckflosse** [comment]
@agriggio I think it's a matter of refactoring the code.

**Comment by agriggio** [comment]
@heckflosse great, then I can give it a try :-)

**Comment by heckflosse** [comment]
@agriggio absolutely :+1: 
btw: I don't want to hijack this issue but there are cases in dcraw code where black is already subtracted from maximum and other cases where it's not subtracted ...

Edit: At least if I understand dcraw code correctly, which is not so easy...

**Comment by agriggio** [comment]
@heckflosse that's not hijacking, that's in fact the reason why I want to
make sure thumbnails and editor use the exact same code for this: we want
them to agree, and to fix bugs only once when they arise :-)

**Comment by iliasg** [comment]
> I think this is also a good image to test auto WB : http://www.rawtherapee.com/shared/test_images/nikon_d80_wb.nef

For a start we have to add a camconst item because Dcraw's white Level is innaccurate 

```
    { // Quality B
        "make_model": "Nikon D80",
        "dcraw_matrix": [ 8629,-2410,-883,-9055,16940,2171,-1490,1363,8520 ], // Dcraw.c d65
        "ranges": { "white": 3980 } // 12-bit files.
    },
 
```
But even after correcting WL in camconst.json (BL is always correct at 0) the pink thumbnail remain !!. 

@heckflosse 
As about WB .. shouldn't we use the median calculation in all autoWB instances ?. This will save us from problems with Black or White clipped pixels which shift the result of average calculations.

**Comment by heckflosse** [comment]
@iliasg 

> As about WB .. shouldn't we use the median calculation in all autoWB instances ?. This will save us from problems with Black or White clipped pixels which shift the result of average calculations.

That's worth a try at least :+1:

**Comment by iliasg** [comment]
I experimented a bit with nikon_d80_wb.nef ..
As Agriggio wrote, looks like the thumbs code use it's own scale_colors and by this it's using the "As Shot WB" instead of RT's raw_autoWB. 
But this specific sample happens to be a "UniWB" shot (all WB multipliers manually set to almost equal to 1.0), so RB values which would be over the White Level with  a normal camera WB or autoWB are now kept in_range. 

I changed the relative tag (WB_RGGBLevels) in -Nikon: group from 256 257 257 257 to a typical daylight 514 257 257 362 (i.e. 2.0 1.0 1.0 1.41) and the magenta highlights dissappeared
WB on both manualy on the t-shirt 
![thumbs-wb](https://cloud.githubusercontent.com/assets/15012599/24726409/f38772e6-1a5a-11e7-8c92-118254ec56ab.jpg)
Find the relative raw files in https://drive.google.com/drive/folders/0B0NqktTgc54sRFBWak5WazU3dGc?usp=sharing

**Comment by agriggio** [comment]
So, I gave it a shot... this patch tries to make the thumbnail code use the scaling coefficients and black levels that `RawImageSource` uses, more or less. Refactoring the code to share more seems to be tricky, so I went for the easy route :-)
If my understanding of the code is correct, `RawImage::get_colorsCoeff` should compute the correct coefficients, so that we can then apply the scaling "by hand" instead of calling `RawImage::scale_colors`. Note that this doesn't address the auto white balance issue at all, because for that I think the solution proposed by @heckflosse (i.e. read from the cache) makes sense.

Perhaps the patch is equivalent to the one-liner I posted on pixls that modifies `dcraw.cc` directly, but this feels a bit better to me... although I don't really know :-)

Feedback highly welcome!

```diff
diff --git a/rtengine/rtthumbnail.cc b/rtengine/rtthumbnail.cc
--- a/rtengine/rtthumbnail.cc
+++ b/rtengine/rtthumbnail.cc
@@ -58,6 +58,20 @@
         return raw_image.get_thumbOffset() + length <= raw_image.get_file()->size;
     }
 
+
+void scale_colors(rtengine::RawImage *ri, float scale_mul[4], float cblack[4])
+{
+    DCraw::dcrawImage_t image = ri->get_image();
+    int size = ri->get_iheight() * ri->get_iwidth();
+
+    for (int i = 0; i < size * 4; ++i) {
+        float val = reinterpret_cast<unsigned short *>(image)[i];
+        val -= cblack[i & 3];
+        val *= scale_mul[i & 3];
+        reinterpret_cast<unsigned short *>(image)[i] = rtengine::CLIP(val);
+    }
+}
+
 }
 
 extern Options options;
@@ -333,17 +347,23 @@
     tpp->greenMultiplier = ri->get_pre_mul(1);
     tpp->blueMultiplier = ri->get_pre_mul(2);
 
-    ri->scale_colors();
+    //ri->scale_colors();
+    float pre_mul[4], scale_mul[4], cblack[4];
+    ri->get_colorsCoeff(pre_mul, scale_mul, cblack, false);
+    scale_colors(ri, scale_mul, cblack);
+    
     ri->pre_interpolate();
 
     rml.exifBase = ri->get_exifBase();
     rml.ciffBase = ri->get_ciffBase();
     rml.ciffLength = ri->get_ciffLen();
 
-    tpp->camwbRed = tpp->redMultiplier / ri->get_pre_mul(0);
-    tpp->camwbGreen = tpp->greenMultiplier / ri->get_pre_mul(1);
-    tpp->camwbBlue = tpp->blueMultiplier / ri->get_pre_mul(2);
-    tpp->defGain = 1.0 / min(ri->get_pre_mul(0), ri->get_pre_mul(1), ri->get_pre_mul(2));
+    tpp->camwbRed = tpp->redMultiplier / pre_mul[0]; //ri->get_pre_mul(0);
+    tpp->camwbGreen = tpp->greenMultiplier / pre_mul[1]; //ri->get_pre_mul(1);
+    tpp->camwbBlue = tpp->blueMultiplier / pre_mul[2]; //ri->get_pre_mul(2);
+    //tpp->defGain = 1.0 / min(ri->get_pre_mul(0), ri->get_pre_mul(1), ri->get_pre_mul(2));
+    tpp->defGain = max(scale_mul[0], scale_mul[1], scale_mul[2], scale_mul[3]) / min(scale_mul[0], scale_mul[1], scale_mul[2], scale_mul[3]);
+
     tpp->gammaCorrected = true;
 
     unsigned filter = ri->get_filters();
```

**Comment by Floessie** [comment]
Here are two implementations for `scale_colors()` (should be `scaleColors()`) that don't need a `reinterpret_cast` (*shiver*):

```c++
void scale_colors(rtengine::RawImage *ri, float scale_mul[4], float cblack[4])
{
    DCraw::dcrawImage_t image = ri->get_image();
    const int size = ri->get_iheight() * ri->get_iwidth();
#if 1
    for (int i = 0; i < size * 4; ++i) {
        float val = image[0][i];
        val -= cblack[i & 3];
        val *= scale_mul[i & 3];
        image[0][i] = rtengine::CLIP(val);
    }
#else
    for (int i = 0; i < size; ++i) {
        for (int j = 0; j < 4; ++j) {
            float val = image[i][j];
            val -= cblack[j];
            val *= scale_mul[j];
            image[i][j] = rtengine::CLIP(val);
        }
    }
#endif
}
```

The first one yields smaller code, the second one is more accurate.

Best,
Flössie

**Comment by agriggio** [comment]
Thanks @Floessie :+1: , I've integreated your second version.

Btw:
> (should be `scaleColors()`)

No, it shouldn't. In fact, `writingCamelCaseForFunctionNames` should be considered illegal, sinful, almost as bad as not wrapping lines to 80 characters :smile: :stuck_out_tongue_closed_eyes: :stuck_out_tongue_winking_eye:

**Comment by heckflosse** [comment]
@agriggio May I push a 3x speedup for scale_colors for bayer files or shall I post a patch?

**Comment by agriggio** [comment]
Go ahead! no need to ask :-)

**Comment by heckflosse** [comment]
Ok, pushed. I placed a BENCHFUN in scale_colors so you can verify. In my test processing time for scale_colors went down from 180 ms to 55 ms. Something similar is also possible for xtrans files.
Edit: of course it blows up the code a bit...

**Comment by heckflosse** [comment]
@agriggio Do you have unpushed changes on hold or can I try to speedup the xtrans case as well?

**Comment by agriggio** [comment]
go ahead!

**Comment by heckflosse** [comment]
@agriggio I pushed the speedup for xtrans files

**Comment by heckflosse** [comment]
@agriggio As dcraw::scale_colors() is not used anymore, what do you think about this patch?
```
diff --git a/rtengine/dcraw.cc b/rtengine/dcraw.cc
index e55c025..eca3733 100644
--- a/rtengine/dcraw.cc
+++ b/rtengine/dcraw.cc
@@ -4526,6 +4526,7 @@ void CLASS colorcheck()
 //  free (fimg);
 //}
 
+/*
 void CLASS scale_colors()
 {
   unsigned bottom, right, size, row, col, ur, uc, i, x, y, c, sum[8];
@@ -4641,6 +4642,7 @@ skip_block: ;
   }
 }
 
+*/
 void CLASS pre_interpolate()
 {
   ushort (*img)[4];
diff --git a/rtengine/dcraw.h b/rtengine/dcraw.h
index c0502ae..419dae1 100644
--- a/rtengine/dcraw.h
+++ b/rtengine/dcraw.h
@@ -421,7 +421,7 @@ void pseudoinverse (double (*in)[3], double (*out)[3], int size);
 void cam_xyz_coeff (float rgb_cam[3][4], double cam_xyz[4][3]);
 //void hat_transform (float *temp, float *base, int st, int size, int sc);
 //void wavelet_denoise();
-void scale_colors();
+//void scale_colors();
 void pre_interpolate();
 //void border_interpolate (int border);
 //void median_filter();
diff --git a/rtengine/rawimage.h b/rtengine/rawimage.h
index cc6dccb..c704c88 100644
--- a/rtengine/rawimage.h
+++ b/rtengine/rawimage.h
@@ -173,12 +173,12 @@ public:
 
     void getRgbCam (float rgbcam[3][4]);
     void getXtransMatrix ( int xtransMatrix[6][6]);
-    void clearXtransCblack( )
-    {
-        for(int c = 0; c < 4; c++) {
-            cblack[c] = 0;
-        }
-    }
+//    void clearXtransCblack( )
+//    {
+//        for(int c = 0; c < 4; c++) {
+//            cblack[c] = 0;
+//        }
+//    }
     unsigned get_filters() const
     {
         return filters;
@@ -312,14 +312,14 @@ public:
 
 public:
     // dcraw functions
-    void scale_colors()
-    {
-        if(isXtrans()) {
-            clearXtransCblack( );
-        }
-
-        DCraw::scale_colors();
-    }
+//    void scale_colors()
+//    {
+//        if(isXtrans()) {
+//            clearXtransCblack( );
+//        }
+//
+//        DCraw::scale_colors();
+//    }
     void pre_interpolate()
     {
         DCraw::pre_interpolate();

```

**Comment by agriggio** [comment]
:+1:

**Comment by heckflosse** [comment]
@agriggio I removed dcraw::scale_colors() and pushed

I think that's a good point to merge your branch because it solves a serious issue with thumbnail rendering and also the generation of the thumbs from raw is faster than in den branch now. Do you agree?

We can continue with improvements to AutoWB then.

Ingo

Edit: of course we have to disable the stopwatches before merge

**Comment by agriggio** [comment]
:+1: for me, I'll merge in a few moments

**Comment by agriggio** [comment]
merged
